### ISSUE TITLE
Mesas generate without clay with Biomes O Plenty installed

### ISSUE THREAD

**Issue by FlashFyreMods** [SpongePowered/SpongeForge#1834]
When running the latest Biomes O Plenty with the latest Spongeforge, all mesa biomes generate with only stone, and no clay whatsoever. 

It is **definitely** Sponge, as mesas generate normally when installing BOP on its own.

**I am currently running...**
- SpongeForge version: spongeforge-1.10.2-2477-5.2.0-BETA-2698
- Forge version: Forge 1.10.2 12.18.3.2477 (But has been replicated on an older version with _see below_ )
- Java version: Online hosting service, so I don't know (but I have replicated it with Java 8 on _see below_ )
- Operating System: Online hosting service, so I don't know (but has been replicated on Win 7)

**What is the expected result?**
Mesas to generate normally.
**What is the current result?**
Mesas generate like this: 
![unknown](https://user-images.githubusercontent.com/33395370/32415737-17cf2f30-c236-11e7-89bd-6e648a060a97.png)
(There are other mods on the client, hence the minimap, but just BOP and Sponge on the server)

FML log:
[fml-server-latest.log](https://github.com/SpongePowered/SpongeForge/files/1444130/fml-server-latest.log)

To sum up: BOP + Sponge = Messed up mesa biomes with no clay generating

**Comment by Inf3rnado** [comment]
Current server:
Java: Java 8
OS: CentOS 7

**Comment by ryantheleach** [comment]
I'm going to check this out on Sponge 1.12 What seed did you use?

**Comment by ryantheleach** [comment]
This is reproducible on 1.12 `/biomesop tpbiome 37`
[07:38:17] [main/INFO]: [CHAT] Teleported Player475 to biome Mesa at (-9486.0, 68.0, 1012.0)
`-3934076723076037690` World Type: Biomes O' Plenty

This only occurs when using the World Type: Biomes O' Plenty, even with BOP installed.

SpongeCommon class `MesaBiomeGenerationPopulator` is never hit when WorldType is BOP, but is when world type is default.

**Comment by ryantheleach** [comment]
@Deamon5550 
It's failing because BoP is still using the SpongeGenerationPopulator, but isn't running BiomeMesa::genTerrainBlocks because we cancel it, as we handle it in our populator. MixinBiomeMesa::genTerrainBlocks is the mixin that cancels it.

Commenting out the cancel makes it work, but I'm not sure what side effects that has as I havn't investigated fully.

**Comment by ryantheleach** [comment]
When BoP is installed, SpongeForge assumes that the mod will handle it's own generation, so we don't add to it.

![image](https://user-images.githubusercontent.com/141554/32996396-005b9f30-cdd2-11e7-842e-0bbc9a04b17d.png)


However, we have changed the base assumption of where the clay gets populated from `BiomeMesa::genTerrainBlocks` to `MesaBiomeGenerationPopulator` and cancel all calls to `genTerrainBlocks` for BiomeMesa.

genTerrainBlocks is cancelled, even if a mod is using BiomeMesa without possibly knowing about our custom populator.

We either need to 
1. come up with a way of conditionally cancelling if a mod is attempting to treat it like the vanilla class ( I initially thought that I could cancel if the mod was removed from the pop list, but then it would be impossible to remove).
2. alway add the mesa populator as a default unless explicitly removed, or 
3. return this to the way vanilla does it.

@Deamon5550 Any ideas for a choice of strategy?

**Comment by ryantheleach** [comment]
Rushed because I needed to get ready for work, but you can see a preliminary fix here with really dirty git history.

https://github.com/RTLSponge/SpongeAPI/tree/MesaFix
https://github.com/RTLSponge/SpongeCommon/commits/fix/issue-1530-DirtySharing

**Comment by Deamon5550** [comment]
That is not the correct fix, the simple fix should be to make the cancellation of that method conditional on the generation for that world not being mod overriden. Adding a populator while assuming that a mod is handling it will break the generation if a mod is say replacing the mesa generation with their own handling.

**Comment by FlashFyreMods** [comment]
Is it looking promising? Obviously my server would suck if all mesas were like this, and I can't delay the release forever, yknow, i'd like to get things done asap, but i'm an impatient person, so try things at your own pace lol. I plan to backup the map when it's finished, so if a fix arrives after release for whatever reason I can just revert the world. Obviously I know that the parts of mesas which have been already generated will stay as stone, but that isn't a big price to pay.  I don't really understand any of this :stuck_out_tongue_closed_eyes: but it sounds like a fix would be possible _one way or another_, right?

**Comment by ryantheleach** [comment]
@Deamon5550 How do you suggest telling the difference between it being mod generated vs plugin generated? Can I just have an initialization boolean or method that is only set in a different constructor or something?

**Comment by BillyGalbreath** [comment]
Any word on this?

**Comment by Zidane** [comment]
As of the latest SpongeForge build via this common commit:

https://github.com/SpongePowered/SpongeCommon/commit/ad4859eeff55a9f883a93cb43c3b6657a4f40e4a

This is resolved.
